### PR TITLE
Initial Raspberry Pi 3 support

### PR DESCRIPTION
**Pull request by rn**
This adds initial support for Raspberry Pi Model 3b to LinuxKit. 

This is WIP and still requires:
- ~~`moby` tool changes: https://github.com/moby/tool/pull/166~~
- ~~Update of the `moby` tool hash~~
- ~~rebuilding of the kernels~~
I'll do this once I get a LGTM

With this PR you can to a `moby build -format rpi3 foo.yml` on a arm64 Linux machine and get a tar ball which can be extracted onto a SD card from which a RPi3 can be booted. Serial console, video console and networking should all work out of the box. There are some limitations, see the documentation included in this PR.

The commits in this PR:
- Change the kernel config to include the minimum necessary support for the RPi3b (see the commit message for more details)
- Include the device tree binaries (dtbs) in the `kernel.tar` file under `./boot/dtb`
- Add `mkimage-rpi3`, used by the `moby` tool to set up `u-boot`. `mkimage-rpi3` compiles a known good version of `u-boot` and downloads the RPi3 binary firmware blobs.
- Add documentation

wip #2574 

baby-hippo

### CODE REVIEW

File: docs/platform-rpi3.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+console with a USB keyboard should work. As mentioned above, there is
+not graphics support.
+
+
+## Disks
+
+The SD card is used to boot. It should be mountable as a disk in
+LinuxKit but is not really suitable for persistent storage.
+
+TODO: Work out how to use a second partition on the SD card for storage.
```
Comment (hairyhenderson) [inline docs/platform-rpi3.md]: For this part I'd been playing with `sfdisk` a little as in https://github.com/hypriot/image-builder-raw/blob/master/builder/rpi/build.sh#L33..L40

I couldn't get this to fully work, mind you, and I didn't hack on this for long. But it works for HypriotOS, so should be viable.

**Comment by justincormack** [comment]
Merged the `tool` PR.

**Comment by rn** [comment]
Updated the moby tool and the kernels so this PR now also contains updates to kernels released the past days so that we now have the kernel config changes picked up there.

The 4.13.x kernel update was problematic and I had to drop some patches for VMBus stability to conflicts with upstream changes. I hadn't had a change yet to test if this breaks things, but we should merge irrespective as the VMBus patches were not fully tested on 4.13 anyway.

**Comment by rn** [comment]
Hmm, `linuxkit.build.outputs.iso-bios` and `linuxkit.platforms.qemu.run_iso` failed with the ISO build for x86 with `Error writing iso-bios output: docker run failed: exit status 125 output:` but nothing obvious changed in this space in this PR. Running this locally works fine...

Hitting rebuild

**Comment by rn** [comment]
Hmm, failed on the same tests again

**Comment by rn** [comment]
ah, `moby` is trying to use `linuxkit/mkimage-iso-bios:65b051322578cb0c2a4f16253b20f7d2797a502`, which doesn't:
```
docker pull linuxkit/mkimage-iso-bios:65b051322578cb0c2a4f16253b20f7d2797a502
Error response from daemon: manifest for linuxkit/mkimage-iso-bios:65b051322578cb0c2a4f16253b20f7d2797a502 not found
```

The current tag should be:
```
linuxkit pkg show-tag tools/mkimage-iso-bios
linuxkit/mkimage-iso-bios:165b051322578cb0c2a4f16253b20f7d2797a502
```
and it is on hub and it looks like a6b89f113787 ("Update linuxkit/mkimage-*") may have updated it to the wrong hash. Fixing...

**Comment by justincormack** [comment]
Merged the `tool` PR...

**Comment by rn** [comment]
updated the moby commit hash and force pushed. Let's see if this makes the CI green

**Comment by justincormack** [comment]
only gcp failed - hit rebuild

**Comment by justincormack** [comment]
hmm, failed again

**Comment by rn** [comment]
@talex5 @dave-tucker any idea what:
```
"The resource 'projects/moby-datakit-ci/zones/europe-west1-b/instances/linuxkit-ci-6' was not found">
```
could be on this PR at: https://linuxkit.datakit.ci/log/saved/linuxkit-test-gcp-of-acfecefbe25f0370e4754681e80623f2d786cb6b280b3c3cadd0a7b94fd27518/75f66b9a7570249528bdf72b0514051b1e17f1b9

Anyway, I'll hit rebuild again

**Comment by dave-tucker** [comment]
@rn `Exception: {u'errors': [{u'message': u'The tar archive is not a valid image.', u'code': u'INVALID_IMAGE_TAR'}]}` is the problem. The GCP image wasn't able to be created as the `tar` file appears to be in an invalid format.

**Comment by rn** [comment]
@dave-tucker thanks for spotting this...I missed it...looking into it

**Comment by hairyhenderson** [comment]
Not sure if it's relevant, but when I tried to extract the tar created with this branch using the stock macOS `tar` (`bsdtar` 2.8.3), it vomited:

```
$ tar xvf ~/Downloads/minimal.tar  
x ./: Can't create '.'
x ./bootcode.bin
x ./fixup_cd.dat
x ./fixup_x.dat
x ./start.elf
x ./u-boot.bin[1]    99143 segmentation fault  tar xvf ~/Downloads/minimal.tar
```

When I used GNU tar, however, it worked (`gtar` in my installation)... ¯\\\_(ツ)_/¯

**Comment by rn** [comment]
@hairyhenderson works fine on my mac...
```
rn-mbp:rs/rneugeba/tmp/foo: tar xvf ../demo.tar
x ./
x ./initrd.uimg
x ./uboot.log
x ./kernel.uimg
x ./boot.scr
x ./bcm2837-rpi-3-b.dtb
x ./config.txt
x ./boot.script
x ./start_cd.elf
x ./start_x.elf
x ./fixup_x.dat
x ./fixup_cd.dat
x ./start.elf
x ./bootcode.bin
x ./u-boot.bin
x ./fixup.dat
```

**Comment by rn** [comment]
I had some local issues, but was able to build a local tar.gz for gcp and it boots fine on qemu. The `mkimage-gcp` has not changed, still used the same `linuxkit/guestfs` image and only the packaging for it changed. A mystery.

**Comment by rn** [comment]
weird, I just build, pushed and run a linuxkit image on GCP without issues. @dave-tucker or @talex5 I don't have access to the account we use for CI. Could you try the same with it?

**Comment by talex5** [comment]
The most recent failure was failing to pull the PR branch from GitHub. Looks like GitHub notified the CI that d2f2a5c was the new head commit, but when the CI pulled the branch, the commit was no longer there. That could happen if:

1. You pushed d2f2a5c
2. You force-pushed something else
3. The CI tried and failed to get d2f2a5c
4. You force-pushed it back to d2f2a5c

Anyway, I hit rebuild on the fetch step and it worked the second time. It's now building...

**Comment by talex5** [comment]
This does seem like an unusual way to start a tgz file... a print statement going to the wrong stream, perhaps?

```
$ hexdump -C linuxkit-ci-2.tar.gz | head
00000000  54 61 67 67 69 6e 67 20  6c 69 6e 75 78 6b 69 74  |Tagging linuxkit|
00000010  2f 6d 6b 69 6d 61 67 65  2d 67 63 70 40 73 68 61  |/mkimage-gcp@sha|
00000020  32 35 36 3a 64 39 35 37  31 61 35 35 37 65 34 62  |256:d9571a557e4b|
00000030  38 32 61 39 34 34 66 31  32 30 38 32 63 64 35 30  |82a944f12082cd50|
00000040  39 38 37 64 33 37 32 36  33 38 35 62 35 34 35 38  |987d3726385b5458|
00000050  38 34 36 63 62 61 65 38  39 65 61 39 62 64 36 39  |846cbae89ea9bd69|
00000060  34 63 38 35 20 61 73 20  6c 69 6e 75 78 6b 69 74  |4c85 as linuxkit|
00000070  2f 6d 6b 69 6d 61 67 65  2d 67 63 70 3a 64 31 38  |/mkimage-gcp:d18|
00000080  38 33 38 30 39 64 32 31  32 63 65 30 34 38 66 36  |83809d212ce048f6|
00000090  30 62 65 62 30 33 30 38  61 34 64 32 62 31 34 63  |0beb0308a4d2b14c|
```

**Comment by rn** [comment]
Ah, interesting. I think I’ve seen this before. Thanks this is a good pointer

**Comment by rn** [comment]
Force pushed an update with the hopefully fixed `moby` tool (https://github.com/moby/tool/pull/171). Hopefully this makes the CI go green

**Comment by ernoaapa** [comment]
In the description you mention documentation, but I don't find it anywhere. Is it somewhere available?

**Comment by rn** [comment]
yeah, there was a commit in the original PR but it accidentally got dropped when I was frantically force pushing updates to get this through CI. I plan to write them again soon

**Pull request merged**